runtime/src/main/scala/com/soundcloud/twinagle/ServerEndpointBuilder.scala

```diff
@@ -7,8 +7,11 @@ import scalapb.{GeneratedMessage, GeneratedMessageCompanion, Message}
 
 class ServerEndpointBuilder(extension: EndpointMetadata => Filter.TypeAgnostic) {
   // TODO: refactor types to aliases?
-  def build[Req <: GeneratedMessage with Message[Req] : GeneratedMessageCompanion, Rep <: GeneratedMessage with Message[Rep] : GeneratedMessageCompanion](rpc: Req => Future[Rep], endpoint: EndpointMetadata): (EndpointMetadata, Service[Request, Response]) = {
-    val httpService: Service[Request, Response] = new ServiceAdapter(rpc)
-    endpoint -> extension(endpoint).toFilter.andThen(httpService)
+  def build[Req <: GeneratedMessage with Message[Req] : GeneratedMessageCompanion, Resp <: GeneratedMessage with Message[Resp] : GeneratedMessageCompanion](rpc: Req => Future[Resp], endpointMetadata: EndpointMetadata): (EndpointMetadata, Service[Request, Response]) = {
+    val httpService =
+      extension(endpointMetadata).toFilter andThen
+        new TracingFilter[Request, Response](endpointMetadata) andThen
```

Comment:
huh. I suspect the error handling bit won't get called: `ServiceAdapter` catches errors and turns them into appropriate HTTP responses. We may need to split it into a part that handles proto/json serialization, and a separate error handling - to HTTP response - filter that goes around at the top of the chain.

Reply:
never mind, I mixed things up: `ServiceAdapter` doesn't do the error handling; that happens in `Server` which wraps multiple endpoints. This is :fine: